drivers/can/can_stm32.c

```diff
@@ -913,157 +810,94 @@ static inline uint32_t can_stm32_filter_to_ext_id(const struct zcan_filter *filt
 		(1U          << CAN_STM32_FIRX_EXT_IDE_POS);
 }
 
-static inline int can_stm32_set_filter(const struct device *dev, const struct zcan_filter *filter,
-				       int *filter_index)
+static inline int can_stm32_set_filter(const struct device *dev, const struct zcan_filter *filter)
 {
 	const struct can_stm32_config *cfg = dev->config;
-	struct can_stm32_data *device_data = dev->data;
+	struct can_stm32_data *data = dev->data;
 	CAN_TypeDef *can = cfg->master_can;
 	uint32_t mask = 0U;
 	uint32_t id = 0U;
-	int filter_id = 0;
-	int filter_index_new = -ENOSPC;
-	int bank_nr;
+	int filter_id = -ENOSPC;
 	int bank_offset = 0;
-	uint32_t bank_bit;
-	int register_demand;
-	enum can_stm32_filter_type filter_type;
-	enum can_stm32_filter_type bank_type;
+	int bank_nr;
 
 	if (cfg->can != cfg->master_can) {
 		/* CAN slave instance: start with offset */
 		bank_offset = CAN_STM32_NUM_FILTER_BANKS;
 	}
 
 	if (filter->id_type == CAN_STANDARD_IDENTIFIER) {
-		id = can_stm32_filter_to_std_id(filter);
-		filter_type = CAN_STM32_FILTER_STANDARD;
-
-		if (filter->id_mask != CAN_STD_ID_MASK) {
-			mask = can_stm32_filter_to_std_mask(filter);
-			filter_type = CAN_STM32_FILTER_STANDARD_MASKED;
+		for (int i = 0; i < CONFIG_CAN_MAX_STD_ID_FILTER; i++) {
+			if (data->rx_cb_std[i] == NULL) {
```

Comment:
This is fine, but could perhaps be optimised slightly by using an atomic. Just an observation, not blocking at all.

Reply:
Do you think we could avoid the mutex-guarded section from which this function is called by using an atomic? I don't think so, as we also change the filter bank registers, which should not be accessed from two threads in parallel.

And if this section is already protected by a mutex, an additional atomic would only be unnecessary additional overhead, as far as I understand.

Reply:
No, the mutex would still be needed.

Reply:
Ok, so if the mutex is still needed, what would be the "optimization" by using an atomic?

Reply:
Likely none, depending on the instructions generated by the compiler. I had not noticed the existing mutex.